### ISSUE TITLE
Implement Truly Parallel Ensemble Layers

### ISSUE THREAD

**Issue by fiorenza2** [pytorch/pytorch#54147]
## 🚀 Feature
Implement a parallel ensemble layer that allows parallelized forward and backward passes through an ensemble of MLPs, as opposed to using a for-loop over individual networks.

## Motivation
Ensembling fully connected networks is still a popular approach to improving generalization; for instance it is commonly used in many state-of-the-art RL algorithms, such as [here](https://papers.nips.cc/paper/2018/hash/3de568f8597b94bda53149c7d7f5958c-Abstract.html), [here](https://arxiv.org/abs/2005.05951), [here](https://arxiv.org/abs/1812.05905), and [here](https://openreview.net/forum?id=AY8zfZm0tDd). Furthermore, ensembling + aleotoric uncertainty estimates are [well-calibrated on out-of-distribution data](https://proceedings.neurips.cc/paper/2019/hash/8558cb408c1d76621371888657d2eb1d-Abstract.html).

At the moment, the most obvious way to implement these networks is to wrap individual networks into a list, and then for-loop over them to get the ensemble prediction. However this is slow, and instead it should be possible to pass a datapoint simultaneously through the entire ensemble. I've implemented a preliminary approach using `torch.baddbmm` and the speedups can be very considerable:
Relative Speedup             |  Direct Comparison
:-------------------------:|:-------------------------:
<img src="https://user-images.githubusercontent.com/15997485/111479874-4d95bd80-8729-11eb-9f12-127968af0c2d.png" alt="drawing" width="300"/>  |  <img src="https://user-images.githubusercontent.com/15997485/111479904-54bccb80-8729-11eb-87d8-f0f74eacbc49.png" alt="drawing" width="300"/>

The results above are for a simple 3 hidden layer MLP with 256 width units trained on MNIST for 1 epoch, with training done on an RTX 2080 Ti.

## Pitch

I propose to implement a parallel ensemble layer.

## Alternatives

Apart from the aforementioned for-loop, it might be possible to do something with grouped convolutional operations, but I've really not looked at this too deeply, i.e., [here](https://stackoverflow.com/questions/58374980/run-multiple-models-of-an-ensemble-in-parallel-with-pytorch/58389075#58389075).

## Additional context

I'm not hugely experienced in OSS development, and already forsee some issues that someone more experienced than me could help out on. For instance, existing operations, such as loss functions and layer initializations wouldn't work natively with my naive implementation; in the above example I wrote custom functions to handle the ensemble cases.

In terms of performance, the two are comparable, and seem reasonable for 1 epoch of MNIST training:
<img src="https://user-images.githubusercontent.com/15997485/111481944-5a1b1580-872b-11eb-8ee9-5d854eb49db1.png" alt="drawing" width="300"/>

**Comment by zou3519** [comment]
@fiorenza2 thank you for the motivation. Do you have a sense of what the API for something like a parallel ensemble layer would look like?

**Comment by fiorenza2** [comment]
Good question! So what I've done so far is basically copy the API for the standard `nn.Linear` layer, which makes it amenable to things like `nn.Sequential`. The only additional argument is the number of members in the ensemble. Also to facilitate this I've got a check in the `forward` to make sure the input data is the right dimension, and puts repeats where necessary (i.e., in the input layer). Here's a snippet of what I have so far:

```
class EnsembleLinear(Module):

    def __init__(self, in_features, out_features, ensemble_size, bias=True):
        super().__init__()
        self.ensemble_size = ensemble_size
        self.in_features = in_features
        self.out_features = out_features
        self.weights = torch.Tensor(ensemble_size, in_features, out_features)
        if bias:
            self.biases = torch.Tensor(ensemble_size, 1, out_features)
        else:
            self.register_parameter('biases', None)
        self.reset_parameters()
    
    def reset_parameters(self):
        for w in self.weights:
            w.transpose_(0, 1)
            init.kaiming_uniform_(w, a=math.sqrt(5))
            w.transpose_(0, 1)

        self.weights = Parameter(self.weights)

        if self.biases is not None:
            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weights[0].T)
            bound = 1 / math.sqrt(fan_in)
            init.uniform_(self.biases, -bound, bound)
            self.biases = Parameter(self.biases)

    def forward(self, input):
        if len(input.shape) == 2:
            input = input.repeat(self.ensemble_size, 1, 1)
        return torch.baddbmm(self.biases, input, self.weights)

    def extra_repr(self) -> str:
        return 'ensemble_size = {}, in_features={}, out_features={}, biases={}'.format(
            self.ensemble_size, self.in_features, self.out_features, self.biases is not None
        )
```

An immediate issue for me is highlighted in the `reset_parameters` method, which has a 'hacky' in-place transpose to correctly apply the parameter initializations. Perhaps there's some way to specify some specific behavior when invoking the initialization API? Another issue I highlighted is also with the loss functions/output activations; for instance, trying to get the Log-SoftMax of the logits is particularly inelegant: `log_softmax = F.log_softmax(logits.transpose(0,2)).transpose(0,2)`, because the data is effectively `N x B x O` v.s. `B x O`, where `N` is ensemble size, `B` is minibatch size, and `O` is output dimensionality. I've not profiled a for-loop of losses over a `chunk`ed output, but I imagine this would undo some of the nice-ness of the parallelization?

**Comment by zou3519** [comment]
@fiorenza2 as you mentioned, one potential design is to have an "Ensemble{X}" version of each nn.Module X in PyTorch's library. I agree that the two items you've identified (initialization and how to express the loss function) are things we'll need to take into account.

We've been experimenting with having a vmap in PyTorch that is similar to [JAX's vmap](https://jax.readthedocs.io/en/latest/jax.html#jax.vmap). If you haven't heard about vmap before, think about it like a python-for loop (or map) except it generates efficient Tensor code. One of the use cases is training over parallel ensemble layers. I wrote up an [example here](https://github.com/Chillee/pt_transformation_experiments/blob/master/ensembling/parallel_train.py) that roughly shows how one would use this. In particular:
- [this solves](https://github.com/Chillee/pt_transformation_experiments/blob/ad5b2ff19989ba3d6bab38169a2f4b287a25e827/ensembling/parallel_train.py#L57-L58) the "loss function" problem you were mentioning.
- It doesn't quite solve the [initialization problem](https://github.com/Chillee/pt_transformation_experiments/blob/ad5b2ff19989ba3d6bab38169a2f4b287a25e827/ensembling/parallel_train.py#L89-L94), but I think to really fix initializations we'd have to refactor PyTorch's NN modules to take in something like the ensemble size, which is something that we're open to.

I'm curious on what your thoughts are on an approach like this.

**Comment by fiorenza2** [comment]
@zou3519 Thanks for sending the example code over! I actually tried building your PyTorch repo from source to run some timing comparisons, but was not successful (it gets to around 50% 😞). Looking at the code though, the approach you outlined makes a lot of sense, and I like that it retains the familiar 'collection of models design'. Furthermore, it would appear more flexible, supporting different architectures to be ensembled in parallel (there could be an architectural diversity argument for improved generalization in model ensembles, such as [this](https://arxiv.org/abs/1910.00370)). Also agreed about the loss function issue being resolved with this method. Regarding initializations, would it be the case that we could simply for-loop over the list and apply the existing functions individually?

My main concern would be that the `vmap` approach seems to have a dependency on the [functional API](https://github.com/pytorch/pytorch/issues/49171), so would this be blocked if that doesn't come through? Furthermore I noticed the current implementation uses a re-coded SGD, would this mean existing optimizers would need refactoring? Thanks!

**Comment by zou3519** [comment]
@fiorenza2 thank you for trying it out! Are you on the PyTorch slack by the way? I have some questions that might be easier communicated in chat form, but we could also continue chatting on this thread.

> I actually tried building your PyTorch repo from source to run some timing comparisons, but was not successful (it gets to around 50% 😞)

The performance is expected for now, under the hood vmap is just running most operations in a for-loop. So for example, for torch.mm, torch.vmap is just running it in a for-loop and performing stack instead of calling torch.bmm.

If you add `torch._C._debug_only_display_vmap_fallback_warnings(True)` to the top of the file that you're running, it'll give some warnings wherever some future work can speed things up.

Out of curiosity, what were you benchmarking? Was it the linear layer or the mnist example?

> Regarding initializations, would it be the case that we could simply for-loop over the list and apply the existing functions individually?

We can't really call the `reset_parameters()` method on modules on a list of weights. But iterating over a list and applying the individual init functions does work. I'm hoping that we can refactor PyTorch modules in a way that we can ask them to apply initializations for some weights that we provide them.

> My main concern would be that the vmap approach seems to have a dependency on the functional API, so would this be blocked if that doesn't come through?

We've been experimenting with various way to get vmap to actually work with NN Modules. The functional API is one such way. If we choose to go down this route we would ship both :).

> Furthermore I noticed the current implementation uses a re-coded SGD, would this mean existing optimizers would need refactoring?

In the current design, I think that existing optimizers would need refactoring. If we can express each optimizer as a function (something like https://github.com/pytorch/pytorch/issues/45242), then it would work out. However, PyTorch optimizers can hold state and there's a question of how do we design this in a way that is easy to use.

**Comment by fiorenza2** [comment]
@zou3519, applied to join the Slack a couple of days ago so very happy to discuss further there once I get access 😄 

> The performance is expected for now, under the hood vmap is just running most operations in a for-loop. So for example, for torch.mm, torch.vmap is just running it in a for-loop and performing stack instead of calling torch.bmm.

Ah apologies I was not clear with this, I couldn't actually build the PyTorch branch you linked, so haven't got any experimental results for it sadly; I'm getting a GCC error, so I think my Cuda (9.1) might be out of date! I'm going to try again after updating.

> In the current design, I think that existing optimizers would need refactoring. If we can express each optimizer as a function (something like #45242), then it would work out. However, PyTorch optimizers can hold state and there's a question of how do we design this in a way that is easy to use.

This makes a lot of sense, and I'm happy to help if possible, especially if it means ensembling becomes signficantly faster 😄

**Comment by amundra15** [comment]
Hey. For my use case, I will have 21 independent models to train. And thus, having such a parallel technique will be quite helpful. Are there any updates on this feature? Or, do I have any other alternative which is not sequential?

**Comment by zou3519** [comment]
@amundra15 could you take a look at https://github.com/pytorch/functorch/blob/main/notebooks/ensembling.ipynb and let us know what you think?

**Comment by amundra15** [comment]
@zou3519 this definitely looks promising. I will have to verify if it works well with my network, given that I make changes on top of the vanilla NeRF model.

Also, I have a conceptual doubt about PyTorch. Since PyTorch calls are asynchronous, does it mean calling the individual networks in a loop would also result in parallel usage of the GPU? (though I don't expect it to be as good as being able to process them in a batch as you do).

**Comment by lorenzozanisi** [comment]
Have there been any further developments on this?

**Comment by Santiago-R** [comment]
> @amundra15 could you take a look at https://github.com/pytorch/functorch/blob/main/notebooks/ensembling.ipynb and let us know what you think?

This notebook does not exist anymore. Seems it has been ported to torch: https://github.com/pytorch/pytorch/blob/main/functorch/notebooks/ensembling.ipynb
